assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/cl/controller.yaml.tmpl

```diff
@@ -271,6 +271,7 @@ storage:
           -A INPUT -p tcp --dport 2380 -j ACCEPT
           -A INPUT -p tcp --dport 2381 -j ACCEPT
           -A INPUT -p tcp --dport 6443 -j ACCEPT
+          -A INPUT -p tcp --dport 7443 -j ACCEPT
```

Comment:
Yes, that should be conditional. Unfortunately we don't distinguish here between the interfaces types. Ideally we would only open 7443 to cluster internal network and leave 6443 open to the world.

Reply:
Why to the internal network? Isn't all cluster traffic going to 6443 and just HAProxy sends it to localhost on port 7443?

Reply:
No, in-cluster traffic will be going either via port 6443 or port 7443, depending on what IP address is advertised into the endpoints of `kubernetes.default.svc`.

Reply:
What? That is not deterministic and at least only one of those ports will be used?

Also, if in-cluster traffic can go to any IP... I'm kind of confused now with the whole picture of the HAProxy setup being used. When would in-cluster traffic change? Only when the pod is down and pod-checkpointer takes over?

Reply:
Well, github is not showing here that the context changed (now this is conditional) :-/

In any case, can you please elaborate on when/why in-cluster traffic can go to either port?

Reply:
So, this is rather not very well known feature of kube-apiserver as far as I know, but I'll try to keep it short. `--advertise-address` flag of `kube-apiserver` defines which IP address will be added as an endpoint to `kubernetes.default.svc` service, where in-cluster traffic to API will be going. It's uniqueness though it's based on IP address and not on IP+port combination. This is why we randomize the IP address to listen on and we set `--advertise-address` to Node IP, to keep it deterministic. If we would randomize the port, we would have to open all the ports in the range on the firewall to get it to work.

So, as uniqueness of `Endpoint` in `kubernetes.default.svc` `Service` is based on IP address, the address with port 6443 will be added first, as this is how bootstrap `kube-apiserver` binds. This port is later replaced by HAProxy. However, if you delete `kubernetes.default.svc` service (`kubectl delete service -n default kubernetes`), then `kube-apiserver` will re-create with existing listening port, which will be `7443` and then the traffic flows directly into the `kube-apiserver` process. The removal should not happen under regular operation, but we should be prepared for it.

This all would be much simpler, if `kube-apiserver` would be capable of advertising endpoints based on IP + port.

Reply:
Oh, I see... but not sure. The service can specify a port, if we do specify it in the service... wouldn't we just remove the need to use 7443 port for in-cluster communication and can rely only on 6443?

In any case, I don't think these tricks should be understood by one person only. If that happens, the code is not written properly with all the comments it needs for other developers to work on it. IMHO, it might be complicated, but we need to be explicit about hacks, limitations and workarounds and it shouldn't be hidden in a solved github comment in some "past" PR (i.e. if we merge with only this conversation as the only guide)

Reply:
I sometimes think I'm following it and sometimes doubt. We can create an issue or add a comment in a following PR explining this, maybe?

Reply:
>Oh, I see... but not sure. The service can specify a port, if we do specify it in the service... wouldn't we just use the port in the service and remove the need to use 7443 port for in-cluster communication and can rely only on 6443?

This is not about the service port, but the endpoints port. The service port is always 443, but it gets translated into endpoint port by kube-proxy.

>In any case, I don't think these tricks should be understood by one person only. If that happens, the code is not written properly with all the comments it needs for other developers to work on it. IMHO, it might be complicated, but we need to be explicit about hacks, limitations and workarounds and it shouldn't be hidden in a solved github comment in some "past" PR (i.e. if we merge with only this conversation as the only guide)

I'll add some documentation.

Reply:
I added some comments into the file to describe this change.